
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../overview_files/">
      
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.14">
    
    
      
        <title>Methodology - sci-rocket</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.342714a4.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#methodology" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="sci-rocket" class="md-header__button md-logo" aria-label="sci-rocket" data-md-component="logo">
      
  <img src="../assets/logo.svg" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            sci-rocket
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Methodology
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5s-1.65.15-2.39.42zM3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29zm.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14zM20.65 7l-1.77 3.79a7.02 7.02 0 0 0-2.38-4.15zm-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29zM12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/odomlab2/sci-rocket" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    odomlab2/sci-rocket
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="sci-rocket" class="md-nav__button md-logo" aria-label="sci-rocket" data-md-component="logo">
      
  <img src="../assets/logo.svg" alt="logo">

    </a>
    sci-rocket
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/odomlab2/sci-rocket" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    odomlab2/sci-rocket
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../config_lsf/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    LSF profile
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../overview_files/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Sample-sheet and barcodes
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Methodology
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Methodology
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#major-steps" class="md-nav__link">
    <span class="md-ellipsis">
      Major steps
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Major steps">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#optional-steps" class="md-nav__link">
    <span class="md-ellipsis">
      Optional steps
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#downstream-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      Downstream analysis
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sample-demultiplexing-without-hashing" class="md-nav__link">
    <span class="md-ellipsis">
      Sample demultiplexing (without hashing)
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#hashing" class="md-nav__link">
    <span class="md-ellipsis">
      Hashing
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#haplotyping-optional-mus-musculus-cross-experiments-only" class="md-nav__link">
    <span class="md-ellipsis">
      Haplotyping (optional; Mus musculus cross-experiments only)
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#output" class="md-nav__link">
    <span class="md-ellipsis">
      Output
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#major-steps" class="md-nav__link">
    <span class="md-ellipsis">
      Major steps
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Major steps">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#optional-steps" class="md-nav__link">
    <span class="md-ellipsis">
      Optional steps
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#downstream-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      Downstream analysis
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sample-demultiplexing-without-hashing" class="md-nav__link">
    <span class="md-ellipsis">
      Sample demultiplexing (without hashing)
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#hashing" class="md-nav__link">
    <span class="md-ellipsis">
      Hashing
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#haplotyping-optional-mus-musculus-cross-experiments-only" class="md-nav__link">
    <span class="md-ellipsis">
      Haplotyping (optional; Mus musculus cross-experiments only)
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#output" class="md-nav__link">
    <span class="md-ellipsis">
      Output
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="methodology">Methodology</h1>
<p>See <a href="https://teichlab.github.io/scg_lib_structs/methods_html/sci-RNA-seq3.html">here</a> for more information on the library generation of the sci-RNA-Seq3 protocol.</p>
<h2 id="major-steps">Major steps</h2>
<ol>
<li>Check for sanity of provided barcodes and sample-sheet.</li>
<li>Converts BCL files to paired-end .fq.gz files with PCR indexes in header (<strong>bcl2fastq</strong>).</li>
<li>Merges multiple sequencing runs (<code>path_bcl</code>) into one experiment-based file (<code>experiment_name</code>).</li>
<li>Splits paired-end .fq.fz files into smaller (evenly-sized) chunks for parallelization (<strong>fastqsplitter</strong>).</li>
<li>Demultiplexing using the supplied sample-specific barcodes (<strong>sci-rocket</strong>).</li>
<li>Finds exact or nearest match for PCR Index #1 (p5), PCR Index #1 (p7), ligation and/or RT barcode (single match with ≤1 hamming distance).</li>
<li>Generates sample-specific .fastq.gz files with corrected R1 sequence (48nt) and added read-names in R2.</li>
<li>Read-pairs without all four matching barcodes are discarded into separate .fastq.gz files with logs detailing which barcode(s) are (non-)matching.</li>
<li>For samples with a specified hashing sheet, additional hashing procedures are performed.</li>
<li>Performs adapter and low-quality base-trimming (<strong>fastp</strong>).</li>
<li>Read-pairs with a mate ≤10nt after trimming are discarded.</li>
<li>Aligns reads to the supplied reference genome and perform cell-barcode/UMI counting (<strong>STARSolo</strong>).</li>
<li>STAR index can be generated based on supplied genome sequences and annotations.</li>
<li>Per gene and cellular barcode, intronic, exonics and UTR-overlapping reads (UMI) are counted and multi-mapping reads are distributed using the <code>EM</code> method.</li>
<li>Generate demultiplexing/alignment overview. (<strong>sci-dash</strong>)</li>
<li>Generates a HTML report with demultiplexing and alignment statistics.</li>
</ol>
<blockquote>
<p>Parallization is performed per experiment_name and split chunk.</p>
</blockquote>
<h3 id="optional-steps">Optional steps</h3>
<ol>
<li>(<em>Mus musculus</em>-only) Haplotype demultiplexing.</li>
<li>Adds haplotype-specific read tags (HP) to the STARSolo BAM files using known haplotype-specific SNPs (<strong>MGP</strong> + <strong>haplotag</strong>).</li>
<li>Generate haplotype-specific read-counts per gene per cell (H1, H2, UA) (<strong>umi_tools</strong>).</li>
</ol>
<h2 id="downstream-analysis">Downstream analysis</h2>
<p>For downstream analysis, we also maintain an R package to analyze results produced by <strong>sci-rocket</strong> called <a href="https://github.com/odomlab2/scir"><strong>scir</strong></a>.</p>
<h2 id="sample-demultiplexing-without-hashing">Sample demultiplexing (without hashing)</h2>
<p><strong>Example of R1 sequence:</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>      @READNAME 1:N:0:CCGTATGATT+AGATGCAACT
</span><span id="__span-0-2"><a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                        |----p7---|+|----p5----|: p5 is reverse-complemented during demuxxing.
</span><span id="__span-0-3"><a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>      ACTTGATTGTCAGAGCTTTGGTATCCTACCAGTT
</span><span id="__span-0-4"><a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
</span><span id="__span-0-5"><a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>      The R1 sequence should adhere to the following scheme:
</span><span id="__span-0-6"><a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>      First 9 or 10nt:  Ligation barcode
</span><span id="__span-0-7"><a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a>      Next 6nt:    Primer
</span><span id="__span-0-8"><a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a>      Next 8nt:    UMI
</span><span id="__span-0-9"><a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>      Last 10nt:   RT Barcode (sample-specific)
</span><span id="__span-0-10"><a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>
</span><span id="__span-0-11"><a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>      Anatomy of R1 (ligation of 10nt):
</span><span id="__span-0-12"><a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>      |ACTTGATTGT| |CAGAGC| |TTTGGTAT| |CCTACCAGTT|
</span><span id="__span-0-13"><a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>      |-LIGATION-| |Primer| |---UMI--| |----RT----|
</span><span id="__span-0-14"><a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>
</span><span id="__span-0-15"><a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>      Anatomy of R1 (ligation of 9nt):
</span><span id="__span-0-16"><a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>      |CTCGTTGAT| |CAGAGC| |TTTGGTAT| |CCTACCAGTT| |T|
</span><span id="__span-0-17"><a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>      |-LIGATION| |Primer| |---UMI--| |----RT----| |.| &lt;- Extra base.
</span><span id="__span-0-18"><a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>
</span><span id="__span-0-19"><a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>      Corrected R1 sequence (48nt):
</span><span id="__span-0-20"><a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>      |CCGTATGATT| |AGTTGCATCT| |CTCGTTGAT| |CCTACCAGTT| |TTTGGTAT|
</span><span id="__span-0-21"><a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>      |----p7----| |----p5----| |-LIGATION-| |----RT----| |---UMI--|
</span></code></pre></div>
<p>For sample-demultiplexing, the following steps are performed:</p>
<ol>
<li>Extracts p5, p7 PCR indexes from the read-name of R1 and ligation, RT and UMI barcodes from sequence of read 1 (R1).</li>
<li>If no match, corrects p5, p7, ligation and/or RT barcode to nearest match (with max. 1nt difference). If multiple close matches, discard read-pair.</li>
<li>For ligation barcodes of 9nt in length, an extra G is added to the ligation sequence as padding to ensure 48nt R1 sequence.</li>
<li>Add the barcodes to the read-names of read 1 and 2:<br />
<code>@READNAME|P5-&lt;p5&gt;-P7-&lt;p7&gt;|&lt;ligation&gt;|&lt;rt&gt;_&lt;UMI&gt;</code></li>
<li>Generate sample-specific paired-end fq.gz files with corrected R1 sequence (48nt) and R2 sequence.</li>
</ol>
<h2 id="hashing">Hashing</h2>
<p>Reads (R2) containing both a polyA signal (AAAA) and a hashing barcode are used to flag reads as hashing-reads. These reads are used for collecting hashing metrics (with their respective R1) and subsequently removed from the analysis.</p>
<p>To flag reads as hashing-reads, we first check for the presence of the polyA signal (AAAA) in R2 (first occurence). If this signal is present, we check for the presence of the hashing barcode in R2 prior to this poly-A signal. It is assumed that the hashing barcodes are 10nt and are (directly) prior to the poly-A signal (5' - 1nt spacer).</p>
<p>If no match is found using the first 10nt (5' poly-A - 1nt spacer) ; we try again against the closest match (hamming distance=1). If no rescued match is found, we search for the presence of any hashing barcode in the entire R2 sequence prior to the poly-A signal.</p>
<p>The following metrics are generated from hashing reads, per cellular barcode / hash barcode combination:</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-1-1"><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>sequencing_name   hash_barcode    cell_barcode    count    n_umi
</span><span id="__span-1-2"><a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a>test            AGGTAGAGCT      F07_D09_LIG98_P01-C08      100      10
</span><span id="__span-1-3"><a id="__codelineno-1-3" name="__codelineno-1-3" href="#__codelineno-1-3"></a>test            ACGTTGAATG      F07_D09_LIG98_P01-C08      200      15
</span></code></pre></div>
<p>These metrics are used to determine the hashing efficiency and to correct for UMI bias in downstream analysis:</p>
<ul>
<li>count: Total number of hashing reads for that specific cell-barcode / hash-barcode combination.</li>
<li>n_umi: Number of unique UMIs for that specific cell-barcode / hash-barcode combination.</li>
</ul>
<h2 id="haplotyping-optional-mus-musculus-cross-experiments-only">Haplotyping (optional; <em>Mus musculus</em> cross-experiments only)</h2>
<p>As optional procedure, <strong>sci-rocket</strong> can be used to further haplotype the sex-chromosome X of the demultiplexed samples, e.g. in the case of mouse F1 cross-hybrids, see <a href="../overview_files/#haplotyping-optional-mus-musculus-cross-experiments-only">here</a> for more information.
This will download (or symlink) the <a href="http://www.sanger.ac.uk/science/data/mouse-genomes-project">MGP</a> database and perform haplotype-specific read-counting using <strong>whatshap</strong> on F1-informative heterozygous SNPs.</p>
<h2 id="output">Output</h2>
<p>The major output files are the following:</p>
<ol>
<li><strong>Sequence and sample-specific fastq file(s)</strong>:</li>
<li><code>{experiment_name}/demux_reads/{sample_name}_R1.fastq.gz</code></li>
<li><code>{experiment_name}/demux_reads/{sample_name}_R2.fastq.gz</code></li>
<li><code>{experiment_name}/demux_reads/{sample_name}_R1_discarded.fastq.gz</code></li>
<li><code>{experiment_name}/demux_reads/{sample_name}_R2_discarded.fastq.gz</code></li>
<li><code>{experiment_name}/demux_reads/log_{sample_name}_discarded_reads.tsv.gz</code></li>
<li><strong>Alignment files</strong>:</li>
<li><code>{experiment_name}/alignment/{sample_name}_{species}_Aligned.sortedByCoord.out.bam/bai</code></li>
<li><code>{experiment_name}/alignment/{sample_name}_{species}_Solo.out/</code></li>
<li><strong>Demultiplexing/alignment overview</strong>:</li>
<li><code>{experiment_name}/sci-dash/</code></li>
<li><strong>Logging and benchmarking</strong>:</li>
<li><code>{experiment_name}/logs/</code></li>
<li><code>{experiment_name}/benchmarking/</code></li>
</ol>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "..", "features": ["content.code.copy"], "search": "../assets/javascripts/workers/search.d50fe291.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.13a4f30d.min.js"></script>
      
    
  </body>
</html>